Zend/zend_partial.c

```diff
@@ -0,0 +1,930 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: krakjoe <krakjoe@php.net>                                   |
+   +----------------------------------------------------------------------+
+*/
+#include "zend.h"
+#include "zend_API.h"
+#include "zend_interfaces.h"
+#include "zend_closures.h"
+#include "zend_partial.h"
+
+typedef struct _zend_partial {
+	zend_object      std;
+	zend_function    prototype;
+	zval             This;
+	zend_function    func;
+	zend_function    trampoline;
+	uint32_t         argc;
+	zval            *argv;
+	zend_array      *named;
+} zend_partial;
+
+static zend_object_handlers zend_partial_handlers;
+
+static zend_arg_info zend_call_magic_arginfo[1];
+
+#define Z_IS_PLACEHOLDER_ARG_P(p) \
+	(Z_TYPE_P(p) == _IS_PLACEHOLDER_ARG)
+
+#define Z_IS_PLACEHOLDER_VARIADIC_P(p) \
+	(Z_TYPE_P(p) == _IS_PLACEHOLDER_VARIADIC)
+
+#define Z_IS_PLACEHOLDER_P(p) \
+	(Z_IS_PLACEHOLDER_ARG_P(p) || Z_IS_PLACEHOLDER_VARIADIC_P(p))
+
+#define Z_IS_NOT_PLACEHOLDER_P(p) \
+	(!Z_IS_PLACEHOLDER_ARG_P(p) && !Z_IS_PLACEHOLDER_VARIADIC_P(p))
+
+#define ZEND_PARTIAL_IS_CALL_TRAMPOLINE(func) \
+	UNEXPECTED(((func)->type == ZEND_USER_FUNCTION) && ((func)->op_array.opcodes == &EG(call_trampoline_op)))
+
+#define ZEND_PARTIAL_FUNC_SIZE(func) \
+	(((func)->type == ZEND_INTERNAL_FUNCTION) ? sizeof(zend_internal_function) : sizeof(zend_op_array))
+
+#define ZEND_PARTIAL_FUNC_FLAG(func, flags) \
+	((func)->common.fn_flags & flags)
+
+#define ZEND_PARTIAL_FUNC_DEL(func, flag) \
+	((func)->common.fn_flags &= ~flag)
+
+#define ZEND_PARTIAL_FUNC_ADD(func, flag) \
+	((func)->common.fn_flags |= flag)
+
+#define ZEND_PARTIAL_CALL_FLAG(partial, flag) \
+	(ZEND_CALL_INFO(partial) & flag)
+
+static zend_always_inline uint32_t zend_partial_signature_size(zend_partial *partial) {
+	uint32_t count = partial->func.common.num_args + partial->argc;
```

Comment:
Hm, why is this num_args + argc? I would have expected just argc plus the possible variadic parameter added below.

Reply:
This can likely be improved, but it's not that simple because argc may be 1 (...) and the signature still needs to be the size of num_args ....

Reply:
I see. Maybe it should be `MAX(num_args, argc)`?

Reply:
Yeah, that seems okay.

Fixed in a1d76e69bf359888b219be05d1098ebd2dfb762b